kettle/stream.py

```diff
@@ -0,0 +1,298 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Receive push events for new builds and upload rows to BigQuery."""
+
+# pylint: disable=invalid-name
+
+from __future__ import print_function
+
+import argparse
+import json
+import os
+import pprint
+import sys
+import traceback
+import time
+
+import multiprocessing.pool
+
+try:
+    from google.cloud import bigquery
+    from google.cloud import pubsub
+    import google.cloud.exceptions
+except ImportError:
+    print('WARNING: unable to load google cloud (test environment?)')
+    traceback.print_exc()
+
+import model
+import make_db
+import make_json
+
+
+def process_changes(results):
+    """Split GCS change events into trivial acks and builds to further process."""
+    acks = []  # pubsub message ids to acknowledge
+    todo = []  # (id, job, build) of builds to grab
+
+    # process results, find finished builds to process
+    for ack_id, message in results:
+        if message.attributes['event_type'] != 'OBJECT_FINALIZE':
+            acks.append(ack_id)
+            continue
+        obj = message.attributes['object_id']
+        if not obj.endswith('/finished.json'):
+            acks.append(ack_id)
+            continue
+        job, build = obj[:-len('/finished.json')].rsplit('/', 1)
+        job = 'gs://%s/%s' % (message.attributes['bucket_id'], job)
+        todo.append((ack_id, job, build))
+
+    return acks, todo
+
+
+def get_started_finished(gcs_client, db, todo):
+    """Download started/finished.json from build dirs in todo."""
+    acks = []
+    build_dirs = []
+    pool = multiprocessing.pool.ThreadPool(16)
+    try:
+        for ack_id, (build_dir, started, finished) in pool.imap_unordered(
+                lambda (ack_id, job, build): (ack_id, gcs_client.get_started_finished(job, build)),
+                todo):
+            # build_dir, started, finished = gcs_client.get_started_finished(job, build)
+            if finished:
+                if not db.insert_build(build_dir, started, finished):
+                    print('already present??')
+                start = time.localtime(started.get('timestamp', 0) if started else 0)
+                print(build_dir, bool(started), bool(finished),
+                      time.strftime('%F %T %Z', start),
+                      finished and finished.get('result'))
+                build_dirs.append(build_dir)
+                acks.append(ack_id)
+            else:
+                print('???', build_dir, started, finished)
```

Comment:
Please provide a more helpful message for the oncall than `'already present??'` and `'???'`